### ISSUE TITLE
Non-web Spring Boot app exits with return code 1

### ISSUE THREAD

**Issue by mikegehard** [spring-projects/spring-boot#5968]
I have built a super simple application that exits with status code 1 when I run `./gradlew bootRun`.

Code can be found here:
https://github.com/mikegehard/kotlinArchiveChannel/commit/8c9dcdad1e557ae41e84442afc16ff2295f43e13

Upon further investigation, it looks like the inclusion of the Boot dev tools is what is causing the issues because when I remove them, the app exits with a status of 0:

https://github.com/mikegehard/kotlinArchiveChannel/commit/3db70cb8c199bfee1c0ff361ab75a68d2de43ad5

**Comment by philwebb** [comment]
@mikegehard Have you tried the same code without Kotlin in the mix?

**Comment by mikegehard** [comment]
Yep. A 1.3.5 boot app downloaded from start.spring.io has the same issue.

**Comment by aldoluciano** [comment]
Isn't this related to https://github.com/spring-projects/spring-boot/issues/5922 ?

**Comment by wilkinsona** [comment]
@aldoluciano No, not really. #5922 was a problem when running an executable jar which uses Spring Boot's `Launcher`. This problem is when using `gradle bootRun` where the `Launcher` isn't involved at all.

**Comment by wilkinsona** [comment]
In a simple, non-web application the main thread is the only non-daemon thread. `Restarter.immediateRestart` causes a `SilentExitException` to be thrown. This is correctly caught and swallowed by `SilentExitExceptionHandler`. Once the main thread has died, there are no non-daemon threads left so the JVM exits. Given that `SilentExitException` was swallowed, I would expect the JVM to exit with 0 but that's not the case.

**Comment by wilkinsona** [comment]
The behaviour I described above is the JVM's standard behaviour:

```java
package com.example;

import java.lang.Thread.UncaughtExceptionHandler;

public class ExceptionHandlingTest {

	public static void main(String[] args) {
		Thread.currentThread().setUncaughtExceptionHandler(new UncaughtExceptionHandler() {

			@Override
			public void uncaughtException(Thread t, Throwable ex) {
				System.out.println("Swallowing " + ex);
			}

		});

		throw new RuntimeException();
	}

}
```

```bash
$ java com.example.ExceptionHandlingTest; echo $?
Swallowing java.lang.RuntimeException
1
```

**Comment by wilkinsona** [comment]
I've just been down an OpenJDK source code rabbit hole. As far as I can tell, once the main thread has died due to an uncaught exception, the only way to get the JVM to exit with 0 is to call `System.exit(0)`. Even if you create one or more non-daemon threads that live beyond the main thread, once those threads have died gracefully, the JVM will still exit with 1.

Arguably there's not much point in using DevTools with an application that naturally runs and then exits. That said, it would have been nice to avoid the strange exit but that doesn't appear to be possible.

@mikegehard What were you hoping to happen by using DevTools in an application that runs and then exits? We could mask the undesirable exit behaviour by having DevTools spawn a non-daemon thread such that it keeps the JVM alive. This would give an opportunity for code changes to trigger a restart which, in some cases at least, I could see being useful with an application that would have naturally run and then exited.

**Comment by mikegehard** [comment]
Thanks for tracking this down!

My goal was to build up an app incrementally starting with minimal
dependencies. Eventually my app added some timers so it didn't start and
quit immediately but when I discovered this behavior I hadn't gotten that
far yet.

I had heard about the dev tools so I figured I'd add them in as a starting
dependency to see if they were useful.

I don't see any benefit in masking this behavior for the reasons you
pointed out.

**Comment by wilkinsona** [comment]
@philwebb Can you please review the fix I've made in https://github.com/spring-projects/spring-boot/commit/13635201ffcad9bb3c81e9a626fb96f490f51764?

**Comment by philwebb** [comment]
Re-opening to remind myself to review.

**Comment by philwebb** [comment]
Looks good to me. I tweaked a few things in 5b97981c87e35876177ee5513fbaedd16b9db9bf.

**Comment by Yona-Appletree** [comment]
@philwebb I'm still seeing this on 2.0.0.M7. Is there a plan to merge this into 2.0.0,  and is there a workaround in the meantime?

**Comment by philwebb** [comment]
@Yona-Appletree The fix was applied to 1.3.6 so it should already be part of the 2.0 codebase. Can you please open a new issue and provide a small sample application that shows the problem?
